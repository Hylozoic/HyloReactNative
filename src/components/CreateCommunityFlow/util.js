export const slugValidatorRegex = /^[0-9a-z-]{2,40}$/
export const invalidSlugMessage = 'URLs must have between 2 and 40 characters, and can only have lower case letters, numbers, and dashes.'
