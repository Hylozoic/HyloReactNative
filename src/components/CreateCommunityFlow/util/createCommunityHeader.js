import React from 'react'
import {
  Text
} from 'react-native'
import { HeaderBackButton } from 'react-navigation'
import { caribbeanGreen, white, white60onCaribbeanGreen } from 'style/colors'

export default function createCommunityHeader (title, navigation) {
  return {
    title: <Text style={styles.headerTitle}>{title}</Text>,
    color: white,
    fontSize: 10,
    headerStyle: {
      backgroundColor: caribbeanGreen,
      borderBottomWidth: 0,
      shadowRadius: 0,
      shadowOffset: {
        height: 0
      },
      elevation: 0
    },
    headerBackgroundColor: caribbeanGreen,
    headerLeft: <HeaderBackButton onPress={() => navigation.goBack()} tintColor={white60onCaribbeanGreen} />
  }
}

const styles = {
  headerTitle: {
    color: white,
    fontSize: 12,
    fontWeight: 'bold'
  }
}
