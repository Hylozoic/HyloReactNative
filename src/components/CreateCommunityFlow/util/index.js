export const slugValidatorRegex = /^[0-9a-z-]{2,40}$/
export const invalidSlugMessage = 'URLs must have between 2 and 40 characters, and can only have lower case letters, numbers, and dashes.'

export function formatDomainWithUrl (communityDomain) {
  if (!communityDomain) return
  let formattedDomain = communityDomain.replace(/hylo\.com\/c\/?/, '')
  if (formattedDomain !== '') {
    formattedDomain = 'hylo.com/c/' + formattedDomain
  }
  return formattedDomain
}

export function removeUrlFromDomain (communityDomain) {
  if (!communityDomain) return
  return communityDomain.replace('hylo.com/c/', '')
}
