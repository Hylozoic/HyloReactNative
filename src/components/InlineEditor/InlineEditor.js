import React from 'react'
import {
  Text,
  View,
  TextInput,
  TouchableOpacity,
  ActivityIndicator
} from 'react-native'
import { withNavigation } from '@react-navigation/compat'
import { trim, isEmpty, get, flow } from 'lodash/fp'
import MaterialIcon from 'react-native-vector-icons/MaterialIcons'
import { htmlEncode } from 'js-htmlencode'
import { rhino30 } from 'style/colors'
import styles from './InlineEditor.styles'
// Mentions
import { MENTION_ENTITY_TYPE } from 'hylo-utils/constants'
import scopedFetchPeopleAutocomplete from 'store/actions/scopedFetchPeopleAutocomplete'
import scopedGetPeopleAutocomplete from 'store/selectors/scopedGetPeopleAutocomplete'
import PersonPickerItemRow from 'screens/ItemChooser/PersonPickerItemRow'
// Topics
import fetchTopicsForCommunityId from 'store/actions/fetchTopicsForCommunityId'
import getTopicsForAutocompleteWithNew from 'store/selectors/getTopicsForAutocompleteWithNew'
import TopicRow from 'screens/TopicList/TopicRow'

export class InlineEditor extends React.PureComponent {
  constructor (props) {
    super(props)
    this.state = {
      isFocused: false,
      pickerType: null
    }
    this.editorInputRef = React.createRef()
  }

  insertTopic = topic => {
    const markup = createTopicTag(topic)
    return this.insertPicked(topic, markup, this.props.onInsertTopic)
  }

  insertMention = person => {
    const markup = createMentionTag(person)
    return this.insertPicked(person, markup)
  }

  insertPicked = (choice, markup, onInsertCallback = undefined) => {
    const value = this.props.value || ''
    const position = get('selection.start', this.state) || 0

    // This will insert the markup at the current cursors position while padding the markup with spaces.
    const firstSlice = value.slice(0, position)
    const secondSlice = value.slice(position)
    const prePadding = firstSlice.length > 0 && !/\s$/.test(firstSlice) ? ' ' : ''
    let newValue = [firstSlice, prePadding, markup, ' '].join('')

    // gets the position directly after the inserted markup
    const newSelectionStart = newValue.length

    // Append the second part of the value (after the cursor)
    newValue += secondSlice

    this.props.onChange(newValue)

    if (onInsertCallback) onInsertCallback([choice])

    // We use a timeout since the onChange needs to propagate
    // the new value change before setting the new selection
    setTimeout(() => {
      this.setState(() => ({
        selection: {
          start: newSelectionStart,
          end: newSelectionStart
        }
      }))
      this.editorInputRef.current.focus()
    }, 100)
  }

  openPersonPicker = () => {
    const { navigation } = this.props
    const screenTitle = 'Mention'
    navigation.navigate('ItemChooserScreen', {
      screenTitle,
      ItemRowComponent: PersonPickerItemRow,
      pickItem: this.insertMention,
      searchPlaceholder: 'Type here to search for people',
      fetchSearchSuggestions: scopedFetchPeopleAutocomplete,
      getSearchSuggestions: scopedGetPeopleAutocomplete(screenTitle)
    })
  }

  openTopicsPicker = () => {
    const { navigation } = this.props
    const screenTitle = 'Pick a Topic'
    navigation.navigate('ItemChooserScreen', {
      screenTitle,
      ItemRowComponent: TopicRow,
      pickItem: this.insertTopic,
      searchPlaceholder: 'Search for a topic by name',
      fetchSearchSuggestions: fetchTopicsForCommunityId(this.props.communityId),
      getSearchSuggestions: getTopicsForAutocompleteWithNew
    })
  }

  handleFocus = () => {
    this.setState(() => ({
      isFocused: true,
      selection: { start: 1, end: 1 }
    }))
    this.props.onFocusToggle && this.props.onFocusToggle(true)
  }

  handleBlur = () => {
    if (isEmpty(trim(this.props.value))) this.props.onChange('')
    this.setState(() => ({ isFocused: false }))
    this.props.onFocusToggle && this.props.onFocusToggle(false)
  }

  handleSelectionChange = ({ nativeEvent: { selection } }) => {
    this.setState(() => ({ selection }))
  }

  handleSubmit = () => {
    this.setState(() => ({ selection: { start: 0, end: 0 } }))
    this.editorInputRef.current.blur()
    this.props.onSubmit(this.props.value)
  }

  render () {
    const {
      placeholder = 'Details',
      editable = true,
      value,
      onChange,
      onSubmit,
      submitting = false,
      style,
      inputStyle
    } = this.props
    const { isFocused } = this.state
    const hitSlop = { top: 7, bottom: 7, left: 7, right: 7 }
    return (
      <View style={[styles.container, style]} onFocus={this.handleFocus} onBlur={this.handleBlur}>
        <View style={styles.textInputAndTools}>
          <TextInput
            multiline
            editable={!!editable && !submitting}
            onChangeText={onChange}
            onSelectionChange={this.handleSelectionChange}
            placeholder={placeholder}
            placeholderTextColor={rhino30}
            style={[styles.textInput, inputStyle]}
            underlineColorAndroid='transparent'
            value={value}
            ref={this.editorInputRef}
          />
          <View style={styles.toolbar}>
            <TouchableOpacity hitSlop={hitSlop} onPress={this.openPersonPicker}>
              <Text style={styles.toolbarButton}>@</Text>
            </TouchableOpacity>
            <TouchableOpacity hitSlop={hitSlop} onPress={this.openTopicsPicker}>
              <Text style={styles.toolbarButton}>#</Text>
            </TouchableOpacity>
          </View>
        </View>
        <SubmitButton style={{...styles.submitButton, display: (onSubmit && value.length > 0) ? 'flex' : 'none'}} submitting={submitting} active={value.length > 0} onSubmit={this.handleSubmit} />
      </View>
    )
  }
}

export default withNavigation(InlineEditor)

export function SubmitButton ({ style, submitting, active, onSubmit }) {
  if (submitting) {
    return <View style={style}><ActivityIndicator /></View>
  } else {
    return (
      <TouchableOpacity hitSlop={{ top: 5, bottom: 10, left: 10, right: 10 }} disabled={!active} onPress={onSubmit}>
        <MaterialIcon name='send' size={26} style={[style, active && styles.activeButton]} />
      </TouchableOpacity>
    )
  }
}

export const createMentionTag = ({ id, name }) =>
  `[${name}:${id}]`

export const createTopicTag = topic =>
  `#${topic.name}`

export const mentionsToHtml = (text) => {
  const re = /\[([^[]+):(\d+)\]/gi
  const replace = `<a href="#" data-entity-type="${MENTION_ENTITY_TYPE}" data-user-id="$2">$1</a>`
  return text.replace(re, replace)
}

export const newLinesToBr = (text) => {
  const re = /[\r\n|\r|\n]/gi
  const replace = '<br>'
  return text.replace(re, replace)
}

export const toHtml = flow([trim, htmlEncode, mentionsToHtml, newLinesToBr])
