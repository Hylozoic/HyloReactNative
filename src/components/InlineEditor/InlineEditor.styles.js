import { azureRadiance, gunsmoke, rhino05, rhino60, white } from 'style/colors'

export default {
  container: {
    display: 'flex',
    flexDirection: 'row',
    backgroundColor: rhino05,
    padding: 8,
    alignItems: 'flex-end',
    justifyContent: 'flex-end'
  },
  textInputAndTools: {
    backgroundColor: white,
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    borderWidth: 1,
    borderRadius: 10,
    borderColor: '#CCC',
    paddingHorizontal: 10,
    paddingVertical: 5,
    maxHeight: 200
  },
  textInput: {
    flexGrow: 1,
    flexShrink: 1,
    margin: 0,
    padding: 0,
    fontSize: 16,
    textAlignVertical: 'top'
  },
  toolbar: {
    alignSelf: 'flex-end',
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignContent: 'flex-end'
  },
  toolbarButton: {
    alignSelf: 'flex-end',
    marginLeft: 10,
    fontSize: 20,
    fontWeight: '700',
    color: rhino60
  },
  submitButton: {
    marginLeft: 10,
    marginBottom: 8,
    color: gunsmoke,
  },
  activeButton: {
    color: azureRadiance
  }
}
