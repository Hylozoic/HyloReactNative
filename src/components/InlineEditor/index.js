export { default, toHtml } from './InlineEditor'
