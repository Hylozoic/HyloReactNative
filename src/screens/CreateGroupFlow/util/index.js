export const slugValidatorRegex = /^[0-9a-z-]{2,40}$/
export const invalidSlugMessage = 'URLs must have between 2 and 40 characters, and can only have lower case letters, numbers, and dashes.'

export function formatDomainWithUrl (groupDomain) {
  if (!groupDomain) return
  let formattedDomain = groupDomain.replace(/hylo\.com\/groups\/?/, '')
  if (formattedDomain !== '') {
    formattedDomain = 'hylo.com/groups/' + formattedDomain
  }
  return formattedDomain
}

export function removeDomainFromURL (groupDomain) {
  if (!groupDomain) return
  return groupDomain.replace('hylo.com/groups/', '')
}
