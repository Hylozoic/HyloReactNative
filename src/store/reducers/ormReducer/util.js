/**
 * @providesModule store/reducers/ormReducer/util
 */

export function showMessagesBadge (session) {
  // it is not strictly correct to do this every time you receive a message,
  // because the message you're receiving could be in a thread that was already
  // unseen. but since the UI doesn't show the count, just a badge, it just
  // needs to know whether there are more than 0 unseen threads, and this takes
  // care of that.
  //
  // we have to check that the current user exists, because this code could be
  // run while they are logged out.
  const me = session.Me.first()
  if (me) me.increment('unseenThreadCount')
}
