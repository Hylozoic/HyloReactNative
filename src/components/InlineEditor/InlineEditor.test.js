import React from 'react'
import ReactShallowRenderer from 'react-test-renderer/shallow'
import TestRenderer from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react-native'
import { createMockStore, ReactNativeTestingLibraryRoot } from 'util/testing'
import InlineEditor, {
  SubmitButton,
  mentionsToHTML,
  createMentionTag,
  createTopicTag,
  toHTML
} from './InlineEditor'

const props = {
  onChange: jest.fn(),
  onSubmit: jest.fn(),
  value: 'some value',
  placeholder: 'Placeholder Text',
  groupId: 10,
  route: {
    params: {}
  },
  navigation: {
    setParams: jest.fn(),
    getParam: jest.fn()
  }
}

const renderComponent = async (testProps = {}) => {
  return render(
    <ReactNativeTestingLibraryRoot>
      <InlineEditor {...props} {...testProps} />
    </ReactNativeTestingLibraryRoot>
  )
}

describe('mentions and topics', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  it('renders as expected', async () => {
    const { getByPlaceholderText, getByDisplayValue } = await renderComponent()
    expect(await getByPlaceholderText('Placeholder Text')).toBeTruthy()
    expect(await getByDisplayValue('some value')).toBeTruthy()
  })

  it('onChange is called', async () => {
    const onChangeMock = jest.fn()
    const { getByDisplayValue } = await renderComponent({ onChange: onChangeMock })
    const textInput = await getByDisplayValue(props.value)
    const newValue = 'new text value'
    fireEvent.changeText(textInput, newValue)
    expect(onChangeMock).toHaveBeenCalledWith(newValue)
  })

  it('onSubmit is called', async () => {
    const onSubmitMock = jest.fn()
    const { getByTestId } = await renderComponent({ onSubmit: onSubmitMock })
    await fireEvent.press(await getByTestId('submitButton'))
    expect(onSubmitMock).toHaveBeenCalledWith(props.value)
  })

  // it('inserts the markup for a mention', async () => {
  //   let currentValue
  //   const onChangeMock = jest.fn(value => { currentValue = value })
  //   const { getByTestId } = await renderComponent({ onChange: onChangeMock })
  //   // const instance = TestRenderer.create(<InlineEditor {...props} />).getInstance()
  //   // instance.handleSelectionChange({ nativeEvent: { selection: { start: 5, end: 5 } } })
  //   // instance.insertMention({ id: 333, name: 'sdfdfz' })
  //   await fireEvent.press(await getByTestId('mentionPicker'))
  //   expect(onChangeMock).toHaveBeenCalledWith('some [sdfdfz:333] text')
  // })

  // it('inserts the markup for a topic', () => {
  //   const onInsertTopic = jest.fn()
  //   const instance = TestRenderer.create(<InlineEditor {...props} onInsertTopic={onInsertTopic} />).getInstance()
  //   const topic = { id: 333, name: 'sdfdfz' }
  //   instance.insertTopic(topic)
  //   expect(props.onChange).toHaveBeenCalledWith('#sdfdfz some text')
  // })

  // it('handleFocus', () => {
  //   const instance = TestRenderer.create(<InlineEditor {...props} />).getInstance()
  //   instance.handleFocus()
  //   expect(instance.state.isFocused).toBeTruthy()
  // })

  // it('handleBlur', () => {
  //   const newProps = {
  //     ...props,
  //     value: '   '
  //   }
  //   const instance = TestRenderer.create(<InlineEditor {...newProps} />).getInstance()
  //   instance.setState({ isFocused: true })
  //   instance.handleBlur()
  //   expect(instance.state.isFocused).toBeFalsy()
  //   expect(props.onChange).toHaveBeenCalledWith('')
  // })
})

describe('SubmitButton', () => {
  it('renders SubmitButton when submitting', () => {
    const renderer = new ReactShallowRenderer()
    renderer.render(<SubmitButton submitting />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})

it('toHTML', () => {
  expect(toHTML(
    "hello world [tom](3344) [:5] #adlkjdf here's"
  )).toBe(
    '<p>hello world <a href="#" data-entity-type="mention" data-user-id="3344">tom</a> [:5] #adlkjdf here&#39;s</p>\n'
  )
})

it('createTopicTag', () => {
  expect(createTopicTag(
    { id: 333, name: 'topic' }
  )).toEqual(
    '#topic'
  )
})

it('createMentionTag', () => {
  expect(createMentionTag(
    { id: 30, name: 'tom' }
  )).toEqual(
    '[tom](30)'
  )
})

it('mentionsToHTML', () => {
  expect(mentionsToHTML(
    'hello [tom](333) [two](233) world [three](3332)'
  )).toEqual(
    'hello <a href="#" data-entity-type="mention" data-user-id="333">tom</a> <a href="#" data-entity-type="mention" data-user-id="233">two</a> world <a href="#" data-entity-type="mention" data-user-id="3332">three</a>'
  )
})
