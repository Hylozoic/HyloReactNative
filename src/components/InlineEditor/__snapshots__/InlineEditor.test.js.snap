// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SubmitButton when submitting 1`] = `
<View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`renders as expected 1`] = `
<View>
  <View>
    <TextInput
      allowFontScaling={true}
      blurOnSubmit={false}
      editable={true}
      multiline={true}
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onContentSizeChange={[Function]}
      onFocus={[Function]}
      onSelectionChange={[Function]}
      placeholder="Place Holder"
      placeholderTextColor="#c0c6cd"
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value="some text"
    />
    <SubmitButton
      active={false}
      handleSubmit={[Function]}
      submitting={false}
    />
  </View>
</View>
`;
